org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/MavenBuildSupport.java

```diff
@@ -48,16 +77,75 @@ public void update(IProject project, boolean force, IProgressMonitor monitor) th
 		Path pomPath = project.getFile("pom.xml").getLocation().toFile().toPath();
 		if (digestStore.updateDigest(pomPath) || force) {
 			JavaLanguageServerPlugin.logInfo("Starting Maven update for " + project.getName());
-			//TODO collect dependent projects and update them as well? i.e in case a parent project was modified
-			MavenUpdateRequest request = new MavenUpdateRequest(project, MavenPlugin.getMavenConfiguration().isOffline(), true);
-			configurationManager.updateProjectConfiguration(request, monitor);
+			if (shouldCollectProjects()) {
+				IProject[] projects = collectProjects(project, monitor);
+				MavenUpdateRequest request = new MavenUpdateRequest(projects, MavenPlugin.getMavenConfiguration().isOffline(), true);
+				((ProjectConfigurationManager) configurationManager).updateProjectConfiguration(request, true, true, monitor);
+			} else {
+				MavenUpdateRequest request = new MavenUpdateRequest(project, MavenPlugin.getMavenConfiguration().isOffline(), true);
+				configurationManager.updateProjectConfiguration(request, monitor);
+			}
 		}
 	}
 
+	public IProject[] collectProjects(IProject project, IProgressMonitor monitor) {
```

Comment:
not sure about this approach. that'll trigger a full build of the project and its children IIUC.
I'd go with:
1- check if packaging is pom, else bail
2- load project facade
3- collect all modules from mavenproject (does it include those from  active profiles?)
4- for each module: compute path to module, see if matches an open IProject
5- add all found Maven projects. If packaging pom is found, see 1

Reply:
The caveat is that'd would not cover the case where the project is not a module of its parent pom

Reply:
I have used DefaultMaven.collectProjects(List<MavenProject>, List<File>, MavenExecutionRequest). I think it won't trigger a full build. See DefaultMaven.doExecute(MavenExecutionRequest). 

I have tested the spring-boot project. MavenBuildSupport.collectProjects(IProject, IProgressMonitor) takes about 15 seconds, a full build more than three minutes.

I have also checked the algorithm you proposed.

```
public void collectProjects(Collection<IProject> projects, IProject project, IProgressMonitor monitor) {
	if (!ProjectUtils.isMavenProject(project)) {
		return;
	}
	projects.add(project);
	IMavenProjectFacade projectFacade = registry.create(project, monitor);
	if ("pom".equals(projectFacade.getPackaging())) {
		List<String> modules = projectFacade.getMavenProjectModules();
		for (String module : modules) {
			IPath pomPath = ResourcesPlugin.getWorkspace().getRoot().getFullPath().append(module).append("pom.xml");
			IFile pom = ResourcesPlugin.getWorkspace().getRoot().getFile(pomPath);
			IProject p = pom.getProject();
			if (p.isOpen()) {
				collectProjects(projects, p, monitor);
			}
		}
	}
}
```

Both of the methods return 173 projects.
Which method do you advice?

Reply:
15sec is for which implem?
Also your pomPath seems dubious. It will search for modules from the root of the workspace, not from the root of the project

Reply:
> 15sec is for which implem?

MavenBuildSupport.collectProjects(IProject, IProgressMonitor)

> Also your pomPath seems dubious. It will search for modules from the root of the workspace, not from the root of the project

I have checked that.
("/spring-boot-build/spring-boot-project/pom.xml").getProject() returns P/spring-boot-build
("/spring-boot-project/pom.xml").getProject() returns P/spring-boot-project

Reply:
ok so what time for collectProjects(Collection<IProject> projects, IProject project, IProgressMonitor monitor)?

Reply:
1-2 seconds

Reply:
2s vs 15s, there's no debate

Reply:
keep your commit around (another branch) just in case and update the PR to use the faster version